### PR TITLE
Support for no. of chunks smaller than no. of partitions.

### PR DESCRIPTION
**Pull request by kylasa**
Following changes are made with this PR.
1. Code changes for handling no. of chunks smaller than no. of partitions
2. Adding new test cases, which were previously deleted, for no. of chunks smaller than no. of partitions.
3. Also adding test cases, where multiple partitions are handled by a single process.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [X] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [X] All changes have test coverage
- [X] Code is well-documented
- [X] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tools/distpartitioning/data_shuffle.py [hunk 1]
```diff
@@ -436,14 +439,58 @@ def exchange_feature(
         assert np.all(global_eids == data[constants.GLOBAL_EID][idx1])
         partid_slice = id_lookup.get_partition_ids(global_dst_nids)
 
+    # determine the shape of the feature-data
+    # this is needed to so that ranks where feature-data is not present
+    # should use the correct shape for sending the padded vector.
+    # exchange length here.
+    feat_dim_len = 0
+    if featdata_key != None:
```
Comment (Rhett-Ying) [inline tools/distpartitioning/data_shuffle.py:447]: `is not None`

File: tools/distpartitioning/data_shuffle.py [hunk 2]
```diff
@@ -436,14 +439,58 @@ def exchange_feature(
...
+        feat_dim_len = len(featdata_key.shape)
+    all_lens = allgather_sizes(
+        [feat_dim_len], world_size, num_parts, return_sizes=True
+    )
+    if all_lens[0] <= 0:
```
Comment (Rhett-Ying) [inline tools/distpartitioning/data_shuffle.py:452]: any testcase to cover this scenario?

File: tools/distpartitioning/data_shuffle.py [hunk 3]
```diff
@@ -436,14 +439,58 @@ def exchange_feature(
...
+
     for idx in range(world_size):
         cond = partid_slice == (idx + local_part_id * world_size)
         gids_per_partid = gids_feat[cond]
         tids_per_partid = tids_feat[cond]
         local_idx_partid = local_idx[cond]
 
         if gids_per_partid.shape[0] == 0:
-            feats_per_rank.append(torch.empty((0, 1), dtype=torch.float))
+            # feats_per_rank.append(torch.empty((0, 1), dtype=torch.float))
```
Comment (Rhett-Ying) [inline tools/distpartitioning/data_shuffle.py:485]: remove it

**Review by Rhett-Ying — APPROVED** [review]

File: tools/distpartitioning/data_shuffle.py (hunk 2 above)
Comment (kylasa) [reply tools/distpartitioning/data_shuffle.py:452]: Since we support no. of chunks < no. of partitions, we will now run into situations when reading dataset from the disk some processes (when handling more than one partition per process) will not have any data in partitions other than '0'. In these cases, we use a sentinal value of node and edge_feature_tids as [(0,0)] to indicate that there is no data in these cases. In such cases we will hit this code, as pointed above. 

The existing test cases, where no. of partitions more than one per process will cover this corner case.

File: tools/distpartitioning/data_shuffle.py (hunk 1 above)
Comment (kylasa) [reply tools/distpartitioning/data_shuffle.py:447]: Done.

File: tools/distpartitioning/data_shuffle.py (hunk 3 above)
Comment (kylasa) [reply tools/distpartitioning/data_shuffle.py:485]: Done.
